### PR TITLE
Update cy_build.py to support pip and develop mode on Mac.

### PR DESCRIPTION
**Pull request by mdpearson**
Hello,

I'm a software developer at Invitae, and we've been using pysam for quite some time. Recently I saw an update on the 0.8.5-dev branch that supports linking to pysam .so files on Mac. This made it possible for us to port an internal library to Mac. In the course of that project, I modified cy_build.py to:

a. work with pip (the current implementation depends on easy_install's egg placement)
b. work with develop mode (either `python setup.py develop` or `pip install -e`)
c. use the `@rpath` flag within the clang linker's `install_name` so shared libraries do not contain the full path of the virtual environment, if any, in which they were built.

I hope these changes are helpful, please reach out to me if you have any questions, suggestions for improvement, etc. Keep up the great work.

### CODE REVIEW

File: pysam/__init__.py [hunk 1]
```diff
@@ -43,9 +43,26 @@
...
-            os.path.join(dirname, 'include', 'samtools')]
+
+    #
+    # Header files may be stored in different relative locations
+    # depending on installation mode (e.g., `python setup.py install`,
+    # `python setup.py develop`. The first entry in each list is
+    # where develop-mode headers can be found.
+    #
+    htslib_possibilities = [os.path.join(dirname, '..', 'htslib'),
+                            os.path.join(dirname, 'include', 'htslib')]
```
Comment (jmarshall) [inline pysam/__init__.py]: For htslib, code should be using `#include <htslib/foo.h>`, so the correct directory to add to an include path is the parent of a _htslib_ directory, not that directory itself.

Comment (mdpearson) [inline pysam/__init__.py]: Turns out, in my first attempt to support develop-mode I did exactly as you suggested. Somewhat confusingly, though, _htslib_ header files are stored in a subdirectory called `htslib`, twice... For example, when you run `python setup.py develop`, htslib headers can be found in `<$PKG_ROOT/htslib/htslib/hts.h>`, while when you run `python setup.py install` they live in, e.g., `$YOUR_VE/lib/python2.7/site-packages/pysam/include/htslib/htslib/hts.h`.

**Comment by AndreasHeger** [comment]
Thanks, @mdpearson, that is great! I will try to get travis OSX to run and then merge.

**Comment by AndreasHeger** [comment]
Ok, OSX and python on travis is not going to happen..., see https://docs.travis-ci.com/user/languages/python

**Comment by kyleabeauchamp** [comment]
AFAIK, OSX + Travis could be achieved by using the `miniconda` approach that is used by bioconda:

https://github.com/bioconda/bioconda-recipes/blob/master/scripts/travis-setup.sh

**Comment by AndreasHeger** [comment]
Thanks, I will give it a go - do I need to change the language from python to C? This seems to be necessary in all the workarounds I have seen.

**Comment by kyleabeauchamp** [comment]
I think so:

https://github.com/bioconda/bioconda-recipes/blob/master/.travis.yml

**Comment by bioinformed** [comment]
@AndreasHeger @mdpearson This PR looks extremely promising.  Since I do most of my development on OSX, I'm happy to test as well.

File: pysam/__init__.py (hunk 1 above)
Comment (jmarshall) [inline pysam/__init__.py]: Fair enough, so the parent of the _htslib_ directory happens to be called _htslib_ itself.  Not sure that that makes for the ideal `include` directory under _site-packages/pysam_, but this is not my domain…

**Pull request merged**